### PR TITLE
Increase test coverage

### PR DESCRIPTION
**Pull request by taehoonlee**
This PR is to increase test coverage described in "Requests for Contributions" and currently being developed.

### CODE REVIEW

**Comment by fchollet** [comment]
Great! 👍

File: keras/models.py [hunk 1]
```diff
@@ -123,10 +123,7 @@ def get_json_type(obj):
                 '(see keras.io/optimizers).')
         else:
             f.attrs['training_config'] = json.dumps({
-                'optimizer_config': {
-                    'class_name': model.optimizer.__class__.__name__,
-                    'config': model.optimizer.get_config()
-                },
+                'optimizer_config': model.optimizer,
```
Comment (fchollet) [inline keras/models.py]: An optimizer instance is not serializable (it's a Python object). The previous 4 lines were serializing it. Revert this

Comment (taehoonlee) [inline keras/models.py]: I revert it!

**Comment by fchollet** [comment]
Please let me know you when you think this is ready to be reviewed.

**Comment by taehoonlee** [comment]
Currently, the test coverage has been increased by 3.37% (82.59% -> 85.96%) on Python 2.7 with TF backend. I'll let you know after some issues checked. Thank you!

**Comment by taehoonlee** [comment]
The following is a comparison of the test coverages.

- [Before](https://travis-ci.org/fchollet/keras/builds/235838951): the commit where this PR branched from.
- [After](https://travis-ci.org/fchollet/keras/builds/240201479): the last commit before CNTK is merged.

|   | Before | After |
|:-:|---|---|
| Python 2.7 + TF | 82.59% | 86.76% |
| Python 3.5 + TF | 82.49% | 86.93% |
| Python 2.7 + TH | 80.95% | 84.17% |
| Python 3.5 + TH | 80.44% | 84.39% |

As a new backend is added, it is hard to analyze the improvement of the test coverages. Would you please check this version @fchollet?

File: keras/datasets/cifar.py [hunk 2]
```diff
@@ -30,4 +30,5 @@ def load_batch(fpath, label_key='labels'):
     labels = d[label_key]
 
     data = data.reshape(data.shape[0], 3, 32, 32)
+    labels = labels.reshape(-1, 1)
```
Comment (fchollet) [inline keras/datasets/cifar.py]: This isn't necessary, and may even break some people's code

File: keras/datasets/cifar10.py [hunk 3]
```diff
@@ -19,7 +19,7 @@ def load_data():
     num_train_samples = 50000
 
     x_train = np.zeros((num_train_samples, 3, 32, 32), dtype='uint8')
-    y_train = np.zeros((num_train_samples,), dtype='uint8')
+    y_train = np.zeros((num_train_samples, 1), dtype='uint8')
```
Comment (fchollet) [inline keras/datasets/cifar10.py]: This isn't necessary, and may even break some people's code

**Comment by fchollet** [comment]
Additionally, it seems a number of backend tests are currently failing.

**Comment by taehoonlee** [comment]
I would like to appreciate the feedback. I hope that the following answers may help to alleviate your concerns.

- with `pytest.raises(Exception)`, be more specific than Exception, use a specific subclass. We never raise Exception.
 -ans) Each `Exception` has been replaced by its corresponding subclass.
- There are many changes that don't seem to be related to test coverage. Please don't include them.
 -ans) There are two types of the changes: adding test modules and reducing the total number of lines. I hope you know that the reduction can also improve test coverages because the test coverage is defined as `(1 - M) / T` where `M` and `T` are the number of missing and total lines, respectively.
 -ans) The changes on `cifar.py` and `cifar10.py` where you comment are also trying to reduce the total number of lines. But I reverted them as you pointed out potential damages.
 -ans) In `theano_backend.py`, checking invalid data_format lines have been removed due to the redundancy. By adding `text_utils.py`, I factorized over `imdb.py` and `reuters.py`. The factorization is also able to increase test coverages because the total lines were reduced.
- One import statement per line.
 -ans) The lines have been revised.
- Backend tests have been refactored with the addition of the CNTK backend. Any added test there should follow the current conventions, i.e. it should either include CNTK or make it easy to add CNTK in the near future. Currently the added tests are hard-coded to be only for TF and TH and adding CTNK would not be readily possible.
 -ans) First, all the fails in backend tests have been resolved. Also, I revised some hard-coded tests with `KTF` and `KTH`. But I think it's not bad to leave the rest because the functional coverage of CNTK is not enough to make all the tests work on like TF and TH.

File: keras/datasets/imdb.py [hunk 4]
```diff
@@ -94,38 +84,17 @@ def load_data(path='imdb.npz', num_words=None, skip_top=0,
...
 
 def get_word_index(path='imdb_word_index.json'):
-    """Retrieves the dictionary mapping word indices back to words.
-
-    # Arguments
-        path: where to cache the data (relative to `~/.keras/dataset`).
-
-    # Returns
-        The word index dictionary.
-    """
```
Comment (fchollet) [inline keras/datasets/imdb.py]: You should not remove the docstring, it is potentially useful for generating docs. Also I don't think you need to abstract this function, it's just 3 less lines

File: keras/datasets/reuters.py [hunk 5]
```diff
@@ -78,37 +70,17 @@ def load_data(path='reuters.npz', num_words=None, skip_top=0,
...
 
 def get_word_index(path='reuters_word_index.json'):
-    """Retrieves the dictionary mapping word indices back to words.
-
-    # Arguments
-        path: where to cache the data (relative to `~/.keras/dataset`).
-
-    # Returns
-        The word index dictionary.
-    """
```
Comment (fchollet) [inline keras/datasets/reuters.py]: Same

File: tests/keras/backend/backend_test.py [hunk 6]
```diff
@@ -861,6 +912,164 @@ def test_random_binomial(self):
...
+                        backend.pool3d(x, pool_size=pool_size, pool_mode='median')
+
+    def test_resize_images(self):
+        for data_format in ['channels_first', 'channels_last']:
+            shape = (5, 5)
+            if data_format == 'channels_first':
+                x_shape = (2, 3) + shape
+            elif data_format == 'channels_last':
+                x_shape = (2,) + shape + (3,)
+            check_single_tensor_operation('resize_images', x_shape, BACKENDS, height_factor=2, width_factor=2, data_format=data_format)
```
Comment (fchollet) [inline tests/keras/backend/backend_test.py]: Line too long

File: tests/keras/backend/backend_test.py [hunk 7]
```diff
@@ -861,6 +912,164 @@ def test_random_binomial(self):
...
+                backend.resize_images(x, 2, 2, data_format='channels_middle')
+
+    def test_resize_volumes(self):
+        for data_format in ['channels_first', 'channels_last']:
+            shape = (5, 5, 5)
+            if data_format == 'channels_first':
+                x_shape = (2, 3) + shape
+            elif data_format == 'channels_last':
+                x_shape = (2,) + shape + (3,)
+            check_single_tensor_operation('resize_volumes', x_shape, [KTH, KTF], depth_factor=2, height_factor=2, width_factor=2, data_format=data_format)
```
Comment (fchollet) [inline tests/keras/backend/backend_test.py]: Line too long (and several other lines in this file as well)

File: tests/keras/legacy/interface_test.py [hunk 8]
```diff
@@ -109,6 +110,10 @@ def test_lstm_legacy_interface():
     new_layer = keras.layers.LSTM(2, input_shape=[3, 5], name='d')
     assert json.dumps(old_layer.get_config()) == json.dumps(new_layer.get_config())
 
+    preprocess_weights_for_loading(new_layer,
+                                   [np.random.random(x) for x in [(5, 2), (2, 2), (2,), (5, 2), (2, 2), (2,), (5, 2), (2, 2), (2,), (5, 2), (2, 2), (2,)]],
```
Comment (fchollet) [inline tests/keras/legacy/interface_test.py]: Line too long

File: tests/keras/legacy/interface_test.py [hunk 9]
```diff
@@ -207,6 +212,10 @@ def test_gru_legacy_interface():
     new_layer = keras.layers.GRU(2, input_shape=[3, 5], name='d')
     assert json.dumps(old_layer.get_config()) == json.dumps(new_layer.get_config())
 
+    preprocess_weights_for_loading(new_layer,
+                                   [np.random.random(x) for x in [(5, 2), (2, 2), (2,), (5, 2), (2, 2), (2,), (5, 2), (2, 2), (2,)]],
```
Comment (fchollet) [inline tests/keras/legacy/interface_test.py]: Line too line

**Comment by taehoonlee** [comment]
I agree with the feedback, thank you @fchollet. The lines have been updated.

File: keras/datasets/text_utils.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
+# -*- coding: utf-8 -*-
```
Comment (fchollet) [inline keras/datasets/text_utils.py]: Please move this to preprocessing/sequence

**Comment by taehoonlee** [comment]
Updated.

**Comment by fchollet** [comment]
LGTM, thanks!

**Pull request merged**